### PR TITLE
Improve Validation and Error Handling in QCNN Sequence Method

### PR DESCRIPTION
**Pull request by amansingh2116**
#Resolves https://github.com/SaashaJoshi/piQture/issues/42

This PR enhances the QCNN class's `sequence` method to include more robust validation and error handling for the list of layer operations. The changes aim to prevent incorrect or incompatible layer configurations and improve readability of error messages for users.

### Validation of Layers:
- Ensures that each layer passed to `sequence`:
  - Is a class.
  - Inherits from `BaseLayer`.
  - Is not the abstract `BaseLayer` itself.

### Parameter Checks:
- Verifies that parameters provided for each layer are in a dictionary format.

### Error Messages:
- Improved error messages for clearer troubleshooting.
- Enforced matching of error messages in tests using regex to dynamically verify alignment with messages established in `qcnn.py`.

### Test Coverage:
- Added unit tests to confirm valid and invalid layer inputs:
  - Test for single valid layer and multiple valid layers.
  - Test for rejection of non-layer classes, built-in functions, and invalid parameter types.
  - Enhanced tests to dynamically match expected error messages without hardcoding them in the test suite.

### Parameterization:
- Added parameterized test cases for non-class operations, built-in types (e.g., `dict`, `str`, `object`), and library-specific classes such as `MERA` and `QCNN` to verify that invalid types are appropriately caught.
- Consolidated related test cases into the `test_operation` function for better organization.

---

### Test Suite Consolidation:
- Deleted the standalone `test_qcnn.py` file created earlier for initial testing purposes.
- Integrated its test cases into the existing `tests/test_qcnn.py` file to improve maintainability and consistency.

### New Minimal Tests Added:
- `test_valid_layer`: Verifies that valid layers (subclasses of `BaseLayer`) are processed correctly.
- `test_regular_class_rejected`: Ensures classes not inheriting from `BaseLayer` are rejected.

### Dynamic Error Matching in Tests:
- Updated tests to dynamically match error messages using regex, ensuring alignment with the messages raised in `qcnn.py`. Example:
  ```python
  with raises(TypeError, match=r"Operation at index \d+ must be a class, got .*?"):
      _ = QCNN(num_qubits).sequence(operations)
  ```

---

## Why These Changes Were Made:

1. **Regex Pattern Matching in Tests:**
   - Ensures that the error messages verified in the tests align with the ones raised in `qcnn.py`.
   - Avoids hardcoding new error messages within tests, relying instead on matching patterns dynamically.

2. **Expanded Test Coverage:**
   - Added parameterizations to test edge cases, including:
     - Built-in types like `dict`, `str`, and `object`.
     - Library-specific classes such as `MERA` and `QCNN` to ensure invalid types are properly handled.

3. **Organized Testing:**
   - Consolidated all test cases into a single file to improve clarity and maintainability.

---

## Testing:
- All tests now pass after resolving earlier mismatches.
- Verified against both valid and invalid inputs to ensure correct behavior.
- Comprehensive test coverage ensures robust validation of layer inputs and error handling for the `sequence` method.

### CODE REVIEW

**Comment by SaashaJoshi** [comment]
Hi @amansingh2116, Thanks for the PR. I will get back to this asap.

**Comment by amansingh2116** [comment]
Is there something that I should change or some error while merging the PR?

**Comment by SaashaJoshi** [comment]
Hi @amansingh2116,
The unit tests and linting tests seem to fail for the current code. Can you run `tox -e lint` to see where you can make changes? Let me know if you will need more help here.

**Comment by SaashaJoshi** [comment]
Also, the unit tests are failing because there already exists a `test_qcnn.py` file in the `tests` directory. You will have to merge your code from the test file to the existing `test_qcnn.py` file in order for the github actions to trigger the correct tests.

File: piqture/neural_networks/qcnn.py [hunk 1]
```diff
@@ -84,8 +94,6 @@ def sequence(self, operations: list[tuple[Callable, dict]]) -> QuantumCircuit:
                 unmeasured_bits=unmeasured_bits,
                 **params,
             )
-            # Optionally collect circuit since it is
-            # composed in place.
             _, unmeasured_bits = layer_instance.build_layer()
 
-        return self.circuit
+        return self.circuit
```
[the remark this answers is not in the document]
Comment (amansingh2116) [reply piqture/neural_networks/qcnn.py]: Hi! @SaashaJoshi , I’ve addressed the comments about `test_qcnn.py` & linting issues and updated the error handling logic. Please review the latest changes.

**Comment by amansingh2116** [comment]
I have fixed the linting issues and merged the test cases in the original `test_qcnn.py` file. Please let me know if I should change something else too to merge the PR. @SaashaJoshi

File: tests/neural_networks/test_qcnn.py [hunk 2]
```diff
@@ -82,11 +83,19 @@ def test_operation_tuple(self, num_qubits, operations):
     )
     def test_operation(self, num_qubits, operations):
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:84]:
```suggestion
    def test_operation_isclass(self, num_qubits, operations):
```

File: tests/neural_networks/test_qcnn.py [hunk 3]
```diff
@@ -140,3 +156,39 @@ def test_sequence(self, num_qubits, operations):
...
+        [
+            (2, [(print, {})]),
+            (3, [(sum, {})]),
+            (4, [(lambda x: x, {})]),
+        ],
+    )
+    def test_non_class_operations(self, num_qubits, operations):
+        """Tests that non-class callables are rejected."""
+        with raises(TypeError, match="must be a class"):
+            _ = QCNN(num_qubits).sequence(operations)
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:172]: We are already checking if the operation is a class in the `test_operation` function. You can add the new test cases to the parameterizations in the above function.

File: tests/neural_networks/test_qcnn.py [hunk 4]
```diff
@@ -140,3 +156,39 @@ def test_sequence(self, num_qubits, operations):
...
+
+    @pytest.mark.parametrize(
+        "num_qubits, operations",
+        [
+            (2, [(dict, {})]),
+            (3, [(str, {})]),
+            (4, [(object, {})]),
+        ],
+    )
+    def test_non_baselayer_operations(self, num_qubits, operations):
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:182]: The keyword `operations` refers to the entire dictionary containing various operation/layers. I am renaming the function to `test_non_baselayer_operation`

```suggestion
    def test_non_baselayer_operation(self, num_qubits, operations):
```

File: tests/neural_networks/test_qcnn.py [hunk 5]
```diff
@@ -140,3 +156,39 @@ def test_sequence(self, num_qubits, operations):
...
+        """Tests that non-class callables are rejected."""
+        with raises(TypeError, match="must be a class"):
+            _ = QCNN(num_qubits).sequence(operations)
+
+    @pytest.mark.parametrize(
+        "num_qubits, operations",
+        [
+            (2, [(dict, {})]),
+            (3, [(str, {})]),
+            (4, [(object, {})]),
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:179]: You can also add parameterizations with other classes from inside the library such as as, `MERA`, `QCNN`, etc. to verify if the test case catches them as well.

File: tests/neural_networks/test_qcnn.py [hunk 6]
```diff
@@ -140,3 +156,39 @@ def test_sequence(self, num_qubits, operations):
...
+        ],
+    )
+    def test_non_baselayer_operations(self, num_qubits, operations):
+        """Tests that classes not inheriting from BaseLayer are rejected."""
+        with raises(TypeError, match="must inherit from BaseLayer"):
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:184]: The match argument must contain the exact error message that we are looking for. Use regex here as well.

```suggestion
        with raises(TypeError, match=r"Operation at index \d+\ must inherit from BaseLayer, got \*?"):
```

File: tests/neural_networks/test_qcnn.py [hunk 7]
```diff
@@ -140,3 +156,39 @@ def test_sequence(self, num_qubits, operations):
...
+            _ = QCNN(num_qubits).sequence(operations)
+
+    @pytest.mark.parametrize(
+        "num_qubits, operations",
+        [(2, [(BaseLayer, {})])],
+    )
+    def test_baselayer_itself(self, num_qubits, operations):
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:191]:
```suggestion
    def test_baselayer(self, num_qubits, operations):
```

File: tests/neural_networks/test_qcnn.py [hunk 8]
```diff
@@ -140,3 +156,39 @@ def test_sequence(self, num_qubits, operations):
...
+        """Tests that BaseLayer itself is rejected."""
+        with raises(TypeError, match="cannot be BaseLayer itself"):
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:193]: The match argument must be corrected here as well.

**Review by SaashaJoshi — CHANGES_REQUESTED** [review]
Hey,

The unit tests look very nice now. I am suggested a few changes, mostly dealing with how we can match error messages by utilizing a regex expression in the `raises` class. Thanks for your hard work, really appreciated!

**Comment by amansingh2116** [comment]
I have fixed the Regex pattern matching issue. Please take a look @SaashaJoshi to review the Pull Request.

File: piqture/neural_networks/qcnn.py [hunk 9]
```diff
@@ -12,70 +12,93 @@
...
+
+            if not issubclass(layer, BaseLayer):
+                raise TypeError(
+                    f"Operation at index {idx} must inherit from BaseLayer, got {layer.__name__}"
+                )
+
+            # Prevent using BaseLayer itself
+            if layer is BaseLayer:
+                raise TypeError(
+                    f"Operation at index {idx} cannot be BaseLayer itself, got *"
```
Comment (SaashaJoshi) [inline piqture/neural_networks/qcnn.py:91]: What does the asterisk denote here? We can simply have:
```suggestion
                    f"Operation at index {idx} cannot be BaseLayer itself."
```

File: tests/neural_networks/test_qcnn.py [hunk 10]
```diff
@@ -140,3 +149,33 @@ def test_sequence(self, num_qubits, operations):
...
+
+    @pytest.mark.parametrize(
+        "num_qubits, operations",
+        [(2, [(BaseLayer, {})])],
+    )
+    def test_baselayer(self, num_qubits, operations):
+        """Tests that BaseLayer itself is rejected."""
+        with raises(
+            TypeError,
+            match=r"Operation at index \d+\ cannot be BaseLayer itself, got \*?",
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:179]: Making corresponding changes here.
```suggestion
            match=r"Operation at index \d+\ cannot be BaseLayer itself.",
```

File: tests/neural_networks/test_qcnn.py [hunk 11]
```diff
@@ -140,3 +149,33 @@ def test_sequence(self, num_qubits, operations):
...
+        [
+            (2, [(dict, {})]),
+            (3, [(str, {})]),
+            (4, [(object, {})]),
+            (2, [(MERA, {})]),
+            (3, [(QCNN, {})]),
+        ],
+    )
+    def test_non_baselayer_operation(self, num_qubits, operations):
+        """Tests that classes not inheriting from BaseLayer are rejected."""
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:164]:
```suggestion
        """Tests if the layer operation inherits from the abstract BaseLayer class."""
```

File: tests/neural_networks/test_qcnn.py [hunk 12]
```diff
@@ -140,3 +149,33 @@ def test_sequence(self, num_qubits, operations):
...
+            match=r"Operation at index \d+\ must inherit from BaseLayer, got \*?",
+        ):
+            _ = QCNN(num_qubits).sequence(operations)
+
+    @pytest.mark.parametrize(
+        "num_qubits, operations",
+        [(2, [(BaseLayer, {})])],
+    )
+    def test_baselayer(self, num_qubits, operations):
+        """Tests that BaseLayer itself is rejected."""
```
Comment (SaashaJoshi) [inline tests/neural_networks/test_qcnn.py:176]:
```suggestion
        """Tests if the layer operation is the abstract BaseLayer class."""
```

**Review by SaashaJoshi — COMMENTED** [review]
Hi @amansingh2116,

Thanks for the quick changes. I made a few quick suggestions, apart from that everything lgtm!

**Comment by amansingh2116** [comment]
Done @SaashaJoshi! Thanks for the guidance.

**Comment by SaashaJoshi** [comment]
Thanks for the contribution, @amansingh2116 !
